src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -74,10 +74,13 @@ internal CompletionSource(ITextView textView, IThreadingContext threadingContext
                 _textView.Options.GlobalOptions.SetOptionValue(NonBlockingCompletionEditorOption, true);
             }
 
-            if (!triggerLocation.Snapshot.TextBuffer.Properties.ContainsProperty(PotentialCommitCharacters))
-            {
-                triggerLocation.Snapshot.TextBuffer.Properties.AddProperty(PotentialCommitCharacters, service.GetRules().DefaultCommitCharacters);
-            }
+            var commitCharacters = service.GetRules().DefaultCommitCharacters;
+
+            // In case of calls with multiple completion services for the same view (e.g. TypeScript and C#), we should merge potential commit characters together.
```

Comment:
Still not sure why we're merging: if we're using stale protential commit characters, isn't that a perf problem? We might be doing unnecessary blocking for a language behavior that isn't active anymore?

Reply:
Chatted with Ivan: this is strange because by merging it means the potential commit characters of one language in a projected case is polluting a later session. Either:

1. Merging isn't necessary, and we can just overwrite.
2. Merging is necessary, and we need some sort of cleanup when the session is done.

Reply:
We get stale characters only if this language was present in a view, and then is removed from the view. How often does it happen? If it happens, what's the probability that this language will be re-added to the view?

The potential commit characters prevent Editor from doing semi-expensive mapping (cost is greater than zero, and we are on typing hot path) 

If user types a commit character, Editor then figures out which buffers are available at a given location and calls `ShouldCommitCompletion` on applicable instances of `IAsyncCommitCommitManager`